### DISCUSSION TITLE
utPLSQL Performance Test Suite

### DISCUSSION THREAD

**Discussion by PhilippSalvisberg** [utPLSQL/utPLSQL#1333] [Feature Requests]
### Motivation
With utPLSQL v3.1.4 - v3.1.6 some performance related issues were introduced. See https://github.com/utPLSQL/utPLSQL/issues/882 and https://github.com/utPLSQL/utPLSQL/pull/888. Fixing the issue required some collaboration with others, which worked very well. However, we should introduce a mechanism to avoid or at least try to avoid such problems in the future. 

### Feature Request
I suggest to write performance tests suites with the following main goal:

- compare overall runtime between utPLSQL releases (current / previous)
  - running reporter consumer in the main process (after test completion)
  - running reporter consumer in dedicated processes (during test run)

When running a reporter consumer in a dedicated process the overhead of a reporter (in most cases) should be reduced to something like an `INSERT` in the producing main process. In such a setup a reporter should never be the main contributor to the overall end-to-end runtime.

Technically I think the current development environment could be used for these performance tests. It should be good enough to identify relative performance issues between the current develop branch and the latest release. A dedicated shell script could do the job.

**Comment by jgebal** [comment]
We could do similar setup for suite setup (parsing/reading cache tables)

**Comment by lwasylow** [comment]
There are few areas that could benefit from constant monitoring, I think quite important is to decide at what point it will run and on what volume. I believe this should be part of develop only volume wise not sure what we consider big but not too big.

**Comment by jgebal** [comment]
The only concern on my side is the build time.
Maybe we could do performance testing:
-  as a separate parallel build (on one DB version only?) - next item on matrix
- only as part of release build (only on release branch)
- only to be ran locally (ad-hoc)

Even today I notice that it's quite annoying to have to wait 9-10 minutes for build to finish.

**Comment by jgebal** [comment]
I see three main areas of performance testing:
- cache and parsing
    - initial scanning schema for suites (new schema)
    - consecutive scanning for suites (some package recompiled)
    - running without scanning (read from cache)
- output buffer on large scale of data produced
- coverage gathering on large schema

All three scenarios could be covered with a schema containing significant number of packages and lines of code in spec/body.

We could use a package generator. Here is an old script i used(attached).
[generate_and_check_parsing_for_big_schema.txt](https://github.com/utPLSQL/utPLSQL/files/3031632/generate_and_check_parsing_for_big_schema.txt)
